### DISCUSSION TITLE
A new permission model for pretix

### DISCUSSION THREAD

**Discussion by raphaelm** [pretix/pretix#3461] [Ideas] [answered]
## Review of the current model

Within every organizer account, pretix allows to have create teams:

    Organizer <1----n> Team <n----n> User
                         |
                         |- <n----n> Event 

Every team has a set of permissions, which is a choice of currently 13 boolean flags divided into two groups.

The first group (can_create_events, can_change_teams, can_change_organizer_settings, can_manage_customers,  can_manage_reusable_media, can_manage_gift_cards) work on organizer level. A user who is a member of a team with e.g.  can_manage_customers can manage all customers of this organizer.

The second group (can_change_event_settings, can_change_items, can_view_orders, can_change_orders, can_view_vouchers, can_change_vouchers, can_checkin_orders) works on event level. A team can either grant this permission for the entire organizer account (`all_events = True`) or only for a select set of events (`limit_events`).

We have been very careful to keep the list of boolean flags short and only extend very very slowly.

While this model has served us quite well for many years, there are a few downsides that keep coming up, mostly (but not only) with larger customers.

### Granularity

With only 13 boolean flags, the model lacks granularity. For example, popular use cases that are not possible currently include:

- Allowing users from the finance department to run statistical reports but not see personal data.

- Allowing some users to change some settings of the event (e.g. the event's date and location) but not others (e.g. the payment settings).

- Allowing some users to change the name on a ticket or approve ticket purchases but not make cancellations and issue refunds.

### Extensibility

The model is currently not extensible through plugins. This is a problem, since plugins can introduce major new functionality that warrants separate handling of permissions.

One example is the pretixPOS box office system, which adds a new section to the organizer backend. This section includes various things, among them access to the session reports and individual transactions of the box office. Currently, this view requires the can_change_organizer_settings permission, which is the most powerful permission available, and therefore requires everyone from the accounting department to hold that permission. There's no other permission we could sensible use – can_view_orders comes close, but is an event-dependent permission and not all pretixPOS objects are specific to an event. It would be great if the plugin could add a new permission for this.

This is just one example, there are many more like it out there. Note that adding more extensibility is a superset of adding more granularity, by adding extensibility it will necessarily become more granular as well. So it makes sense to think of these two together instead of individually.

### Forwards-compatibility

This problem [isn't new](https://github.com/pretix/pretix/issues/39), we've had it as an open task since 2015.
The **main** concern why I never tackled it is the following problem:

If we want to have a more fine-grained system, the first step is easy: We add more boolean flags. We go from can_change_orders to stuff like can_cancel_orders, can_approve_orders, can_refund_orders, and so on. We "just" need to add the flags and them use them in the code.

This will easily turn us from 13 flags to 100+. This is fine if we build a good UI around it.

What I'm worried about is what happens next. Suppose in two years, we finally defeat physics introduce the very handy feature of teleportation and we add a new "teleport customer to event" button in the backend and, because teleportation is expensive, we'll add a new permission flag "can_teleport_customer". How would we choose the default value of that flag for all existing teams?

If, in the general case, every team can have any combination of our 100+ flags, the only "safe" way to do this is to set it to "false" for all existing teams. We can send out a newsletter and hope administrators will take the time to go to their teams page and change the flag to true for the teams they want to be able to use this. However, this depends ton administrators being very informed and proactive about even little new features they don't really care about.

I'm worried that this will lead to us spending great effort on new features, and the majority of our users never seeing that they exist! It has the potential to make pretix much less "discoverable" and have every other support call and in "oh, we have that feature, you just haven't assigned yourself the permission yet".

### Backwards-compatibility

If we change the permission model, we'll need both a way to migrate all existing teams to the new model.
To do so, we need a mapping of old permissions to new permissions, for example:

    can_change_orders → [can_cancel_orders, can_approve_orders, can_refund_orders, …]

We will however also need a *safe* way to keep running plugins which have not yet adjusted to the new permission model.
This means a call to `user.has_permission("can_change_orders")` must still return a useful, consistent result even if we moved to a more fine-grained permission model.

## Idea for an approach

I propose the following:

1.) We add a more explicit **categorization** of permissions. A permission can be declared to be related to orders, vouchers, etc.

2.) We add **more fine-grained permissions**, such as the split of can_change_orders mentioned above. In every category, we keep a "catch-all" permission that corresponds to all non-split behaviour of the old permission, to allow for backwards-compatibility.

    Permission(type=EVENT, name="pretix.orders.view", category="orders", label="Can view orders")
    Permission(type=EVENT, name="pretix.orders.cancel", category="orders", label="Can cancel orders")
    Permission(type=EVENT, name="pretix.orders.approve", category="orders", label="Can approive orders")
    Permission(type=EVENT, name="pretix.orders.refund", category="orders", label="Can refund orders")
    Permission(type=EVENT, name="pretix.orders.change", category="orders", label="Can make changes to orders not covered by other permissions")

3.) We add a **mapping table** from old permissions to the new catch-all permission for runtime checks and from old permissions to all new permissions that replaced it for the one-time data migration.

4.) We add a **plugin signal** that allows a plugin to declare a permission. The plugin can either use an existing category or define a new one. For example, a plugin could declare the following:

    Permission(type=ORGANIZER, name="pretix_posbackend.transactions.view", category="pretixPOS", label="Can view transactions")

5.) We add a new concept of **roles**. Roles are pre-defined sets of permissions, hard-coded into pretix. We could add roles like "admin", "event manager", "finance", …, corresponding to typical roles in an organization that uses pretix. We can do a survey of current permission usage to find out typical permission combinations within teams. We will build the UI in a way that encourages the user to select one of these roles, and we'll make it clear that when a new feature is added, **we** decide which roles will get this permission automatically. The roles will basically be a "managed set of permissions". Only power users who really need to full granularity should need to tick checkboxes one by one, and only those who really need this power should carry the burden of monitoring for new features. Plugins will need a way to attach their permissions to roles as well.

I don't think we need the additional complexity of "custom roles" at this time.

To allow working with dynamic permissions, and also to deal with the large number of permissions, we'll probably need to move from a collection of BooleanFields to a JSONField or something like it. We will need to consider how we store a team that "follows" one of the above-mentioned roles, especially when those roles itself are dynamic on their own, due to plugins. I think we might want to build the Team model in a way that it *either* has a role set *or* a list of permissions in a JSON field, but not both.

## Open issues

There's one request we hear from time to time that is not solved by this approach: The idea of per-product permissions. For example, you have a ticket shop that sells both attendee tickets and exhibitor tickets, and you want someone to have access to only the exhibitors tickets. This could mean partial access to the product settings (price, availability), but more often means order data. I see why this would be useful and I have been in situations where I would have liked to have that myself. However, I don't see how to implement it without adding *enormous* levels of complexity (how would you configure that on organizer-level when multiple events have different product structures?) and dealing with lots and lots of really nasty edge cases (if an order includes both products I can see and products I can't see, should I be able to cancel it? See its payment details? change it's invoice address?).

I can only really imagine this to work for read-only access to attendee lists and statistics for a specific product, which would be really useful, but I don't see how to make it fit into the concept as a whole.

**Comment by wiffbi** [comment]
I think this is a reasonable approach! A few things that just came to my mind:

- I think it would be nice to have both, a role and a list of permissions. As soon as a team’s permissions diverge from our standard roles, how do we handle new permissions? We would need to default to false and have the same problem as described above with regards to new features with new permissions. If we allow to selectively overwrite permissions of a role on team level, we would still be able to fallback to a role’s permissions – basically like the organizer settings tickle down to event settings.
- Another thing, that might be helpful: finer grouping than categorization. E.g. a way to allow everything as order_change does currently with one click instead of checking 4 to 5 new checkboxes. This might be an UI issue only (visually grouping and adding group-checkboxes through JS), but we could also use this in a way that allows to set it as a fallback for new sub-permissions … e.g. when setting `pretix.orders.change` to True, it would default to True for all settings named `pretix.orders.change.XXX`, if not set otherwise. That means we should rename `pretix.orders.approve` to `pretix.orders.change.approve`, etc.? That way plugins could add permissions with a more sensible default than True or False, but based on „parent“ permissions. Not sure, this gets too complex though.

**Reply by raphaelm** [reply]
> I think it would be nice to have both, a role and a list of permissions.

Hmm. Isn't that too complex to understand again? Can you only add permissions on top of the role or can you also subtract? I believe this will be hard to build good UI for.

> Another thing, that might be helpful: finer grouping than categorization. E.g. a way to allow everything as order_change does currently with one click instead of checking 4 to 5 new checkboxes. This might be an UI issue only (visually grouping and adding group-checkboxes through JS), but we could also use this in a way that allows to set it as a fallback for new sub-permissions … e.g. when setting pretix.orders.change to True, it would default to True for all settings named pretix.orders.change.XXX, if not set otherwise.

Yes, I believe this would be a counter-proposal to my idea of a `pretix.orders.change` catch-all permission, and I like your idea better.

> That means we should rename pretix.orders.approve to pretix.orders.change.approve, etc.? That way plugins could add permissions with a more sensible default than True or False, but based on „parent“ permissions. Not sure, this gets too complex though.

I believe this parent-child relationship should not be based on the name, but on an explicit link to the parent. This could entirely *replace* the concept of categories, by just having a "this is a child of".

The problem with names is that it does not allow plugins to namespace their permissions properly, leading to possible collisions between plugins.

**Reply by wiffbi** [reply]
> > I think it would be nice to have both, a role and a list of permissions.
> 
> Hmm. Isn't that too complex to understand again? Can you only add permissions on top of the role or can you also subtract? I believe this will be hard to build good UI for.

I would allow to „overwrite“ any permission as in allow or deny – not sure what you mean by „adding“ permissions means; adding as in „new/unknown type of permission“ e.g. when a role does not have a default for a plugin-added permission? I think in that case the permissions should be added to the team directly. Basically a role is just a fallback if not otherwise defined. It could even be seen as the „uber-parent“ if we decide to go the route with the parent-child-relation in permissions you suggested (which I like very much!). UI-wise might be tricky to „unset“ a permission as-in to use the fallback of the parent. A checkbox is noth enough, radio-boxes would get cluncky quickly. What about a dropdown that has the following options:

- allow
- deny
- same as parent (deny)
- same as role (deny)

The text in parentheses „(deny)“ would be dynamic depending on the setting. To help scan the list of settings visually we could add an icon before the dropdown, that represents the selected state (e.g. a green check or a red deny-sign, e.g. `.fa-ban`).

> I believe this parent-child relationship should not be based on the name, but on an explicit link to the parent. This could entirely replace the concept of categories, by just having a "this is a child of".

Yeah, categories would be redundant with this. I like that idea very much! Why not base it on name? Wouldn’t the domain/package-dot-naming help in discovering which permission is a parent/child of which?

> The problem with names is that it does not allow plugins to namespace their permissions properly, leading to possible collisions between plugins.

Could we enforce a namespace on plugins registering new permissions? It would limit plugins to group their permission where at the same time basing the default on some other permission somewhere else in the chain. Not sure how we would allow that – having a permission, that is not explicitly set, being based on a different permission somewhere else in the permission-chain instead of its parent permission. That would probably make it way to complex.

**Reply by raphaelm** [reply]
> I would allow to „overwrite“ any permission as in allow or deny – not sure what you mean by „adding“ permissions means;

Yeah, by adding I meant "allow more than the role"

> What about a dropdown that has the following options: 

That sounds good

> Yeah, categories would be redundant with this. I like that idea very much! Why not base it on name? Wouldn’t the domain/package-dot-naming help in discovering which permission is a parent/child of which?
> …
> Could we enforce a namespace on plugins registering new permissions? 

Hmm, I believe we could base it on the name **or** enforce namespaces, but not both?

I'd very much like to be able to have a plugin permission like "pretix_teleport.orders.change.teleport_person" be a child of the non-plugin permission "pretix.orders.change". Otherwise we'd get the before-mentioned "future-compatibility" only for new features in core, not for new plugins.

**Comment by MattDinnery** [comment]
Just to give you some further use cases of needing granularity:
- Our designer needs to be able to edit the header / colours / etc of the organiser and all events, but should not be able to change any other settings
- Our designer needs to be able to see existing vouchers, and their URLs to be able to generate custom artworks
- Our designer may need to be able to add new vouchers (ie if we give her a list of them we haven't created yet), but will never need to edit nor delete existing ones
- Our ticketing manager & gate manager need to be able to manage devices (adding, editing & deleting, as well as viewing their logs), but should not be able to change any other settings
- Our safety advisor needs to be able to approve a particular set of products (stationary vehicle permit, moving vehicle permit & exceptional vehicle permit), and will need to be able to find those orders by customer name, but should not be able to do anything else with those orders, and, for privacy reasons, should not see any other data (ie email address / phone number / address / etc)

**Comment by raphaelm** [comment]
Your use cases are understandable and I sympathize with them, but we're pretty settled not to go *that* granular in the foreseeable future, i.e.

- there will be no permissions for individual settings
- there will be no "add vouchers but not edit them"
- there will be no permission for individual products

**Reply by MattDinnery** [reply]
I think this needs to be looked at again, as I've seen somebody else asking specifically for device settings to be an explicit permission.

Our one-day volunteer Gate Manager last weekend was given the ability to change all organiser settings, simply so that they could add new devices / show the QR codes for devices we'd pre-loaded but needed to provision. That amount of unrestricted access for such a small & simple task is disproportionate.
Those on our Committee who legitimately need to be able to change organiser settings simply aren't around/available when gate steward volunteers turn up - one is driving a van off-site, another is marshalling procession floats off-site, a third was at the other end of the event site checking on items needed to ensure we would be safe to open, and I was in the control room.

**Comment by raphaelm** [comment, marked as answer]
Finished in https://github.com/pretix/pretix/pull/5728
